src/vuln_analysis/utils/chain_of_calls_retriever_base.py

```diff
@@ -0,0 +1,111 @@
+from abc import abstractmethod, ABC
+from pathlib import Path
+from typing import List, Optional
+
+from langchain_core.documents import Document
+
+from vuln_analysis.utils.dep_tree import DependencyTree, Ecosystem
+from vuln_analysis.utils.functions_parsers.lang_functions_parsers import LanguageFunctionsParser
+
+PARENTS_INDEX = 0
+
+EXCLUSIONS_INDEX = 1
+
+METHOD_EXCLUSIONS_INDEX = 2
+
+from vuln_analysis.logging.loggers_factory import LoggingFactory
+
+logger = LoggingFactory.get_agent_logger(f"morpheus.{__name__}")
+
+def calculate_hashable_string_for_function(function_file_name: str, function_name_to_search: str) -> str:
+    return f"{function_file_name};{function_name_to_search}"
+
+def get_extension_of_file(file_path: str):
+    extension_start = file_path.rfind(".")
+    return file_path[extension_start:]
+
+def is_function_callable(document: Document, language_parser, callee_function_file_name: str) -> bool:
```

Comment:
@tmihalac Can you please add type hint to parameter `language_parser` ?

Reply:
OK